packages/service-library/src/servicelib/archiving_utils.py

```diff
@@ -141,49 +161,81 @@ async def unarchive_dir(
     dir_to_compress: Path,
     destination: Path,
     compress: bool,
     store_relative_path: bool,
-    exclude_patterns: List[str] = None,
+    exclude_patterns: Optional[list[str]] = None,
 ) -> Optional[Exception]:
     try:
         compression = zipfile.ZIP_DEFLATED if compress else zipfile.ZIP_STORED
         with zipfile.ZipFile(
             destination, "w", compression=compression
         ) as zip_file_handler:
-            files_to_compress_generator = _iter_files_to_compress(
-                dir_to_compress, exclude_patterns
+            folder_size_bytes = sum(
+                file.stat().st_size
+                for file in _iter_files_to_compress(dir_to_compress, exclude_patterns)
             )
-            for file_to_add in files_to_compress_generator:
-                try:
-                    file_name_in_archive = (
-                        _strip_directory_from_path(file_to_add, dir_to_compress)
-                        if store_relative_path
-                        else file_to_add
-                    )
-
-                    # because surrogates are not allowed in zip files,
-                    # replacing them will ensure errors will not happen.
-                    escaped_file_name_in_archive = _strip_undecodable_in_path(
-                        file_name_in_archive
-                    )
-
-                    zip_file_handler.write(file_to_add, escaped_file_name_in_archive)
-                except ValueError as value_error:
-                    log.exception("Could write files to archive, please check logs")
-                    raise value_error
+            desc = f"compressing {dir_to_compress} -> {destination}"
+            with tqdm(
+                desc=f"{desc}\n", total=folder_size_bytes, **_TQDM_FILE_OPTIONS
+            ) as pbar:
+                for file_to_add in _iter_files_to_compress(
+                    dir_to_compress, exclude_patterns
+                ):
+                    pbar.set_description(f"{desc}:{file_to_add.name}")
+                    try:
+                        file_name_in_archive = (
+                            _strip_directory_from_path(file_to_add, dir_to_compress)
+                            if store_relative_path
+                            else file_to_add
+                        )
+
+                        # because surrogates are not allowed in zip files,
+                        # replacing them will ensure errors will not happen.
+                        escaped_file_name_in_archive = _strip_undecodable_in_path(
+                            file_name_in_archive
+                        )
+
+                        def _write_with_progress(
+                            zip_write_fct, self, data  # pylint: disable=unused-argument
+                        ):
+                            pbar.update(len(data))
+                            return zip_write_fct(data)
+
+                        # Replace original write() with a wrapper to track progress
```

Comment:
Question: is it necessary to replace the handler at every iteration? Can't you just do it once and pass the `pbar`?

Reply:
refactored. please have a look

Reply:
Very nice. I like it with the context manger!